storyblok/field-plugin
# feat(lib): add new ai prompt action

## What?
- It adds a new action to the library called `promptAI()`.
- It adds this functionality to the demo field plugin.
- It adds an error notification in case of using the action through the Sandbox (not allowed once a space is needed)

## Why?

JIRA: SHAPE-8085

The promptAI action is necessary for the new plugin the team is working on and can also unlock new field plugin integrations in the future.

## How to test? (optional)

To test it out, you will need to:

1. Run the Storyfront locally using the `feat/SHAPE-8085-field-plugin-sdk-ai-prompt-new-action` branch (still being reviewed https://github.com/storyblok/storyfront/pull/6930).

2. Open the `packages/demo/src/components/FieldPluginDemo.tsx` file and specify the `targetOrigin` to `http://localhost:3300` (Storyfront running address):
![image](https://github.com/user-attachments/assets/d211cba8-357a-4bde-a627-661c8bb98adc)

3. [Create a new fieldtype](http://localhost:3300/#/me/plugins) and deploy the field plugin demo (you can skip this step and just use the `io8-prompt-ai` created field type):
```sh
yarn
yarn build
yarn workspace demo build
```
Now, copy the generated bundle file (located at `packages/demo/dist/index.js`) and paste it into the newly created field.

4. Add the new field type to any of your stories.
If everything was configured properly, you should see, in your Visual Editor, something similar to:

![Screenshot 2025-01-13 at 21 05 35](https://github.com/user-attachments/assets/288ea28b-113e-4acb-99c9-e947937ae98e)

#### Prompt AI action

1. Fill the `Ask a question` input with any text you want to resolve using the AI feature.
2. Below, in the select input, select the type of query you want to perform.
3. Click on the `Prompt` button.

**Notes**
- If the selected action type is `translate` a language must be informed (otherwise the call will fail)
- If the selected action type is `adjust-tone` a tone (e.g academic) must be informed (otherwise the call will fail)

**Expected errors**

No space is selected in the Field Plugin Editor:
![Screenshot 2025-01-16 at 18 46 42](https://github.com/user-attachments/assets/e479d56c-0c68-438e-824d-fc51a309d42b)

When trying to use the story as context to prompt the AI but using the Field Plugin Editor:
![Screenshot 2025-01-16 at 18 48 39](https://github.com/user-attachments/assets/1f2c28de-bf3d-4f16-a91d-f75e417c837f)

When using it through the Visual Editor and the AI feature is disabled (in the user's space or in the org):
![Screenshot 2025-01-16 at 18 50 29](https://github.com/user-attachments/assets/4c69cb2c-f019-4964-b4c3-d57f1a99dee8)
![image](https://github.com/user-attachments/assets/f5748e5a-1050-4e0b-987d-4c994916b0b9)

When using it through the Visual Editor and the AI is not supported for the selected language:
![image](https://github.com/user-attachments/assets/3a082aa0-a424-4d0a-bc08-116731d676a8)

When no prompt text is provided (in the Visual Editor or Field Plugin Editor):
![Screenshot 2025-01-16 at 18 57 05](https://github.com/user-attachments/assets/9b87eb87-4d69-4a32-a890-3529dc041cca)

When the prompt text is too long (in the Visual Editor or Field Plugin Editor):
![Screenshot 2025-01-16 at 18 55 45](https://github.com/user-attachments/assets/337a28b7-67cc-4898-a7f2-be5381aa1668)

When the call to the backend fails (it shouldn't happen once there are other checks before calling the endpoint but in case it happens it would look like this):
![Screenshot 2025-01-16 at 19 29 45](https://github.com/user-attachments/assets/46759dfe-9ec7-48f4-904c-9746a2c8d62e)
